### ISSUE TITLE
`unused_async` false positive with `async_fn_in_trait`

### ISSUE THREAD

**Issue by benluelo** [rust-lang/rust-clippy#10459]
### Summary

`async_fn_in_trait` triggers `unused_async` when an implementation doesn't `.await` in the body, and suggests to remove the `async`. This produces broken code, as `async` is required by the trait definition.

### Lint Name

unused_async

### Reproducer

I tried this code:

```rust
#![allow(incomplete_features)]
#![feature(async_fn_in_trait)]
#![warn(clippy::unused_async)]

trait HasAsyncMethod {
    async fn do_something() -> u32;
}

impl HasAsyncMethod for () {
    async fn do_something() -> u32 {
        1
    }
}
```

[playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=e3a2765c6e35061985668dfcee75f7c7)

### Version

```text
rustc 1.69.0-nightly (31f858d9a 2023-02-28)
binary: rustc
commit-hash: 31f858d9a511f24fedb8ed997b28304fec809630
commit-date: 2023-02-28
host: x86_64-unknown-linux-gnu
release: 1.69.0-nightly
LLVM version: 15.0.7
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by benluelo** [comment]
Hitting a similar false positive/suggestion-causes-error with `manual_async_fn` and `#![feature(return_position_impl_trait_in_trait)]`: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=f948f4c776b927e1fa71b3861ac14fcc
